### PR TITLE
Set parameter visibility for InstrumentViewer

### PR DESCRIPTION
**Pull request by darominski**
ILL users found that certain parameters stored in IPFs are very verbose, with multiple lines of text, and that has a severe impact on user experience using the Pick tab in the InstrumentViewer, namely the picking slows down significantly. The proposed solution is to add an attribute to certain parameters stored in IPFs defining whether they should be visible from the InstrumentViewer. This attribute is by default set to 'true'.

**To test:**

1. Load some ILL data, e.g. 093409.nxs of D33 from UniTest/ILL/D33
2. Open the InstrumentViewer and go to the Pick tab, hover over a sensitive element of the instrument.
3. There should be only a handful of parameters printed from D33 component-link (marked as 'Parameters from: D33'), and none longer than one line.
4. Go to the instrument/D33_Parameters.xml and change the 'visible' attribute value from 'false' to 'true' of one of the non-visible parameters, e.g. 'logbook_default_parameters'.
5. Re-load D33 data, and go again to the Pick tab of the InstrumentViewer.
6. The parameters contained in the chosen parameter, e.g. 'logbook_default_parameters', should now be printed.

Part of  #31052  

Release notes added.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by DannyHindson — CHANGES_REQUESTED** [review]
The new feature works well. I've added a couple of minor comments\requests + please can you resolve the merge conflict

File: Framework/Geometry/inc/MantidGeometry/Instrument/Component.h [hunk 1]
```diff
@@ -249,6 +249,32 @@ class MANTID_GEOMETRY_DLL Component : public virtual IComponent {
...
+      } else {
+        param = m_map->get(this, p_name);
+      }
+      if (param != Parameter_sptr()) {
+        return param->visible();
+      } else {
+        return false;
+      }
+    } else {
+      // Not parametrized = return empty vector
```
[the remark this answers is not in the document]
Comment (darominski) [reply Framework/Geometry/inc/MantidGeometry/Instrument/Component.h]: Yes indeed. This comment is now modified to better reflect the logic of returning false boolean when the requested parameter is not defined.

File: Framework/Geometry/src/Instrument/Parameter.cpp [hunk 2]
```diff
@@ -89,6 +91,7 @@ std::shared_ptr<Parameter> ParameterFactory::create(const std::string &className
     throw std::runtime_error("ParameterFactory:" + className + " is not registered.\n");
   p->m_name = name;
   p->m_type = className;
+  p->m_visible = true ? visible == "true" : false;
```
[the remark this answers is not in the document]
Comment (darominski) [reply Framework/Geometry/src/Instrument/Parameter.cpp]: Thank you for spotting that, it is indeed the wrong way around. I have corrected it but the second part of your statement still stands and there is redundancy of setting visibility in creating parameters and through the function `ParameterMap::add`. One or the other way of setting visibility could be removed.

[the remark this answers is not in the document]
Comment (darominski) [reply Framework/Geometry/src/Instrument/Parameter.cpp]: I have removed the redundancy by not setting the visibility in the `ParameterMap` but rather keeping the logic in `Parameter::create ` function.

**Review by DannyHindson — DISMISSED** [review]
Thanks for the changes. Looks good now

**Review by martyngigg — COMMENTED** [review]
I think as this stands it wouldn't survive a save/reload from NeXus if parameter were created outside of the XML in the code with `ParameterMap::create` and `visible="false"`?

The ParameterMap is currently saved to a a processed NeXus file in https://github.com/mantidproject/mantid/blob/4ae5d8cf8a139ffda94f02a8ed8366932373238b/Framework/Geometry/src/Instrument/ParameterMap.cpp#L1016

and read back by https://github.com/mantidproject/mantid/blob/7d925b9547bc98235c739200a876e23509eaf492/Framework/API/src/ExperimentInfo.cpp#L1137

How easy would that be to change?

**Comment by darominski** [comment]
> I think as this stands it wouldn't survive a save/reload from NeXus if parameter were created outside of the XML in the code with `ParameterMap::create` and `visible="false"`?
> 
> The ParameterMap is currently saved to a a processed NeXus file in
> 
> https://github.com/mantidproject/mantid/blob/4ae5d8cf8a139ffda94f02a8ed8366932373238b/Framework/Geometry/src/Instrument/ParameterMap.cpp#L1016
> 
> and read back by
> 
> https://github.com/mantidproject/mantid/blob/7d925b9547bc98235c739200a876e23509eaf492/Framework/API/src/ExperimentInfo.cpp#L1137
> 
> How easy would that be to change?

That is actually an important thing I omitted. Until the fix, all parameters after re-loading would be set to visible regardless of their previous status. The solution is not too complex in the end: one more value is added to the saved NeXus file, with an identifying key: "visible:" followed by the visibility value, either "true" or "false". A simple addition of the value at the end of the list of saved tokens per parameter was not viable (that was my original approach, but I had to scrap it, apologies for the mess), as "true" or "false" may be part of the parameter value for parameters containing multiple semicolons. 

To test: 
1. Load a raw data from an instrument with visibility parameter defined, e.g. `093409.nxs` of D33 from UniTest/ILL/D33
2. Save it and re-load, verify that the same parameters are visible from the InstrumentView for both the originally loaded file and one that went through save-load cycle.
3. Load a file from an instrument that does not have visibility parameters defined, e.g. anything from outside ILL.
4. All parameters defined in the IPF should be visible from the InstrumentView.

**Review by martyngigg — APPROVED** [review]
This looks good. Thanks for making these changes.

**Comment by martyngigg** [comment]
@darominski This was my fault for not checking this but the updates to the IDFs have not been reflected in the schema. Windows clean builds currently check this due to a missing package requirement elsewhere. We had a failure to [validate](https://builds.mantidproject.org/view/Master%20Pipeline/job/master_systemtests-windows/1371/testReport/junit/SystemTests/ValidateInstrumentDir/ValidateParameterFiles/).

Could you look into updating the schema at https://github.com/mantidproject/mantid/blob/master/instrument/Schema/IDF/1.0/IDFSchema.xsd ?

**Comment by darominski** [comment]
> @darominski This was my fault for not checking this but the updates to the IDFs have not been reflected in the schema. Windows clean builds currently check this due to a missing package requirement elsewhere. We had a failure to [validate](https://builds.mantidproject.org/view/Master%20Pipeline/job/master_systemtests-windows/1371/testReport/junit/SystemTests/ValidateInstrumentDir/ValidateParameterFiles/).
> 
> Could you look into updating the schema at https://github.com/mantidproject/mantid/blob/master/instrument/Schema/IDF/1.0/IDFSchema.xsd ?

Thank you for reaching out, I have modified the IPF schema instead of IDF as these set parameters that are intended for being hidden are defined in IPF. The related PR is #32043 .

**Pull request merged**